src/app.jsx

```diff
@@ -55,38 +56,44 @@ export const Application = () => {
     useEffect(() => {
         if (path === undefined)
             return;
-        setFiles([]);
-        const currentPath = path.slice(0, pathIndex).join("/");
-        const channel = cockpit.channel({
-            payload: "fslist1",
-            path: `/${currentPath}`,
-            superuser: "try",
-            watch: true,
-        });
 
-        channel.addEventListener("message", (ev, data) => {
-            const item = JSON.parse(data);
-            if (item.event === 'present' && item.path[0] !== ".")
-                setFiles(f => [...f, { name: item.path, type: item.type, modified: item.modified }]);
-            else if (item.event === 'deleted') {
-                const deleted_name = item.path.slice(item.path.lastIndexOf("/") + 1);
-                setFiles(f => f.filter((res) => { return res.name !== deleted_name }));
-            } else if (item.event === 'created') {
-                const created_name = item.path.slice(item.path.lastIndexOf("/") + 1);
-                const date_modified = item.tag.slice(item.tag.indexOf("-") + 1, item.tag.lastIndexOf("."));
-                if (created_name[0] !== ".") {
-                    setFiles(f => [...f, { name: created_name, type: item.type, modified: date_modified }]);
+        const getFsList = () => {
+            if (channel.current !== null)
+                channel.current.close();
```

Comment:
This is sloppy. Please close the channel immediately in the `ready` message handler, which you will need anyway (see above). Then you also don't need to keep this around as a ref.

Reply:
If I close the channel on ready I will not receive any events

Reply:
Right, sorry, that was bogus; you want the "done" handler, not "ready" (which happens before any events)

Reply:
Sorry, no, doc/protocol.md literally says:

> After all files have been listed the "ready" control message will be sent.

So "ready" is right after all. I think this delineates the time between the "initial read" and the "watch events" (but this isn't using watching)

Reply:
Meh .. it could be that the documentation is wrong. fslist1 isn't being used often, and it's not tested *at all* with the C bridge. But we do have a [Pybridge test for it](https://github.com/cockpit-project/cockpit/blob/main/test/pytest/test_bridge.py#L384), which *does* use `"done"` at the end, not "ready".

Reply:
I see, I thought `getFsList()` would always open a new temporary non-watch fslist1 channel, but it seems you want to make this work with a single watch-enabled channel. So `done` can't work with that of course. If `ready` works as advertised, then feel free to use that and close this thread. If it doesn't, then use a temporary new channel in getFsList(), and if possible move to fswatch1 for the persistent watch channel.

Reply:
ready appears after all 'present' events are received.

Reply:
Nice!